service/src/main/java/org/broadinstitute/listener/relay/inspectors/SamResourceClient.java

```diff
@@ -46,6 +50,14 @@ public Instant checkPermission(String accessToken) {
         var apiClient = getApiClient(accessToken);
         var resourceApi = new ResourcesApi(apiClient);
 
+        // check that user has access to workspace
+        boolean workspaceAccess =
+            resourceApi.resourcePermissionV2("workspace", workspaceId.toString(), "read");
+        if (!workspaceAccess) {
+          logger.error("Unauthorized request. User doesn't have access to workspace.");
+          return Instant.EPOCH;
+        }
```

Comment:
I won't block the PR for this - but it would be great if this was optional and controlled by a value sent by Leo. This extra Sam call is extraneous for `kubernetes-app-shared` apps, e.g. WDS. It won't break anything, but it's an extra API call that could add latency and/or the chance of transient failures.

Reply:
Is this something that would apply statically to the listeners instance? Like, WDS has a listeners instance, and it could safely ignore all workspace access checks?

Reply:
Exactly - the listener deployed for WDS could skip this check; the listener deployed for Cromwell still needs the check. When CBAS moves from single-user to multi-user it could also skip this check.

Reply:
we could tidy two birds with one stone and update leonardo to send a new samChecker parameter `workspaceId`, which if specified would do this check

Reply:
I am ok with calling this out of scope for this bug ticket but I won't speak for WDS as I'm not familiar

Reply:
For WDS would it be setting a config value in Relay Listener that Leonardo will set? I am not very familiar how and when this happens

Reply:
YMMV but,,

You'd probably be making changes around here: https://github.com/DataBiosphere/leonardo/blob/84549cc1f5571f867f087b9e5c39de0b267e9670/http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AKSInterpreter.scala#L169

That's where the configuration is applied to the terra-azure-relay-listeners instance for a new AKS app. You'd want to check if it's a shared app, which I think is what's happening for example here: https://github.com/DataBiosphere/leonardo/blob/84549cc1f5571f867f087b9e5c39de0b267e9670/http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AKSInterpreter.scala#L413

Otherwise you could set the new field in the helmchart values, as well as in AzurePubSubHandler which creates the listener for new runtimes: https://github.com/DataBiosphere/leonardo/blob/84549cc1f5571f867f087b9e5c39de0b267e9670/http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AzurePubsubHandler.scala#L130 (and a corresponding change in azure_vm_init_script)

Reply:
Would it make sense to do that in a separate ticket instead of this ticket? It might require more investigating and changes in more places than Relay Listener

Reply:
Feels ok to me to have this be a first pass with a followup ticket to refine.

Reply:
Is the workspace check actually extraneous for shared apps? What about this scenario:
1. I create a shared app in a workspace
2. I leave the workspace
3. I try to access my shared app

Would we need to check the workspace to effectively block (3)?

Reply:
> Would we need to check the workspace to effectively block (3)

This is exactly the scenario I validated in [AJ-1147](https://broadworkbench.atlassian.net/browse/AJ-1147). Since `kubernetes-app-shared` roles are inherited from `workspace` roles - and ONLY inherited, there are no direct roles - this works as expected. See also https://github.com/DataBiosphere/leonardo/pull/3496

Reply:
Ah I see, cool